File: src/pitono/base_then.py
from typing import Any, Dict, List

class BaseThen:
    def __init__(self, key: str, then_cb: Any):
        self.key = key
        self.then_cb = then_cb
        self.error = False
        self.artifacts: List[str] = []
    
    async def but_then(
        self,
        store: Any,
        then_cb: Any,
        test_resource_configuration,
        pm: Any
    ) -> Any:
        # Execute the then callback which should perform assertions
        # The then_cb is typically a function that takes the store and performs checks
        if callable(then_cb):
            return then_cb(store)
        return True
    
    def add_artifact(self, path: str) -> None:
        normalized_path = path.replace('\\', '/')
        self.artifacts.append(normalized_path)
    
    def to_obj(self) -> Dict[str, Any]:
        return {
            'name': self.key,
            'error': self.error,
            'artifacts': self.artifacts,
            'status': not self.error
        }
    
    async def but_then(
        self,
        store: Any,
        then_cb: Any,
        test_resource_configuration,
        pm: Any
    ) -> Any:
        # Default implementation - can be overridden by subclasses
        return True
    
    async def test(
        self,
        store: Any,
        test_resource_configuration,
        t_log: Any,
        pm: Any,
        filepath: str
    ) -> Any:
        try:
            add_artifact = self.add_artifact
            # Simple implementation for now
            result = await self.but_then(
                store,
                self.then_cb,
                test_resource_configuration,
                pm
            )
            return result
        except Exception as e:
            self.error = True
            raise e

File: src/lib/rubeno/base_then.rb
module Rubeno
  class BaseThen
    attr_accessor :name, :then_cb, :error, :artifacts, :status
    
    def initialize(name, then_cb)
      @name = name
      @then_cb = then_cb
      @error = false
      @artifacts = []
    end
    
    def add_artifact(path)
      normalized_path = path.gsub('\\', '/')
      @artifacts << normalized_path
    end
    
    def but_then(store, then_cb, test_resource_configuration, pm)
      raise NotImplementedError, "but_then must be implemented by subclasses"
    end
    
    def to_obj
      {
        name: @name,
        error: @error,
        artifacts: @artifacts,
        status: @status
      }
    end
    
    def test(store, test_resource_configuration, t_log, pm, filepath)
      begin
        # Call the then_cb with the store to get a boolean result
        result = @then_cb.call(store)
        @status = true
        result
      rescue => e
        @status = false
        @error = true
        raise e
      end
    end
  end
end
